### DISCUSSION TITLE
Missing documentation about configuration syntax/keywords/grammar

### DISCUSSION THREAD

**Discussion by selefotsifem** [fail2ban/fail2ban#3873] [Q&A]
Hopefully someone can get me useful information.
I am looking for the documentation about all the configuration "verbs" and how they are related to each other.
When I browse through several example configuration files, I see a lot of different items, some are explained, others not.
But I miss he big picture with all possibilities and perhaps remarks, which syntax elements are deprecated, recommended, necessary, optional, ... and which item needs other items to complete one configuration task. 

What I have overseen?

**Comment by sebres** [comment]
See https://github.com/fail2ban/fail2ban/issues/2986#issuecomment-811856528.
For anything else your question is too vague, please be more concrete or else provide exact points or examples...

> which syntax elements are deprecated, recommended, necessary, optional, ... and which item needs other items to complete one configuration task.

Depends on what you're calling "configuration task"...
To configure jails, it is enough to create a section with a jail name in `/etc/fail2ban/jail.local` and:
1. for existing jails, add `enabled = true` (and check `backend`, `logpath` or `journalmatch` and few others to ensure the monitoring would be correctly - service writes to journal or log-file, the logfile(s) matches `logpath`, etc)
2. for new jail, you have to specify at least a `filter` or `failregex` (with optionally `prefregex` and/or `datepattern`), proper action (if default action doesn't good for this jai) with other args like `port`, `protocol` etc.

For both one can use `fail2ban-regex` over current log/journal to ensure it matches everything you want.

To write filter or action see [man pages](https://manpages.debian.org/unstable/fail2ban/jail.conf.5.en.html) or [wiki](https://github.com/fail2ban/fail2ban/wiki).

> deprecated

This must be provided in documentation, changelog or related comment in the corresponding configuration (e. g. deprecated actions).

> recommended

It is hard to say, because depends on many factors... But you could read this one - [wiki :: Best practice](https://github.com/fail2ban/fail2ban/wiki/Best-practice).

> necessary, optional

Almost everything is optional, because all parameters have normally default values. 
But it also depends, e. g. some action may need an account name, key, token or password, another - mail address, etc.

> which item needs other items

This is also complex subject and may not be answered generally.

**Comment by selefotsifem** [comment]
Examples, what I am searching:
https://bind9.readthedocs.io/en/v9.18.31/reference.html#
https://nginx.org/en/docs/dirindex.html
https://httpd.apache.org/docs/2.4/ 
  where I can find information in redundant ways, either looking to modules, Configuration Sections, Configuration files, ...m Logfiles, ...
  and finally https://httpd.apache.org/docs/2.4/glossary.html
http://www.squid-cache.org/Doc/config/

Just to name some examples I am used to lookup.
